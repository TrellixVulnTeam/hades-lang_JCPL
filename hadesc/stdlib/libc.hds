
extern def puts(*Byte): Void = puts;

