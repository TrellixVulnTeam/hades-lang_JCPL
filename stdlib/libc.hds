
extern def puts(*Byte): Void = puts;
extern def exit(CInt): Void = exit;
extern def malloc(Size): *mut Void = __hds_malloc;
extern def free(*mut Void): Void = __hds_free;
extern def strlen(*Byte): Size = strlen;
extern def memcpy(*Byte, *Byte, Size): Void = memcpy;
extern def strlen(*Byte): Size = strlen;
/// we don't have variadics right now
/// there isn't a better way to type this
extern def fprintf(*File, *Byte, *Byte): Size = fprintf;
extern def printf(*Byte, *Byte): Size = printf;

struct File {}

