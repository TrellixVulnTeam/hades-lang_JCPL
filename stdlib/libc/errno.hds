type t = Int;

def EPERM(): Int { return  1; }
def ENOENT(): Int { return  2; }
def ESRCH(): Int { return  3; }
def EINTR(): Int { return  4; }
def EIO(): Int { return  5; }
def ENXIO(): Int { return  6; }
def E2BIG(): Int { return  7; }
def ENOEXEC(): Int { return  8; }
def EBADF(): Int { return  9; }
def ECHILD(): Int { return 10; }
def EAGAIN(): Int { return 11; }
def ENOMEM(): Int { return 12; }
def EACCES(): Int { return 13; }
def EFAULT(): Int { return 14; }
def ENOTBLK(): Int { return 15; }
def EBUSY(): Int { return 16; }
def EEXIST(): Int { return 17; }
def EXDEV(): Int { return 18; }
def ENODEV(): Int { return 19; }
def ENOTDIR(): Int { return 20; }
def EISDIR(): Int { return 21; }
def EINVAL(): Int { return 22; }
def ENFILE(): Int { return 23; }
def EMFILE(): Int { return 24; }
def ENOTTY(): Int { return 25; }
def ETXTBSY(): Int { return 26; }
def EFBIG(): Int { return 27; }
def ENOSPC(): Int { return 28; }
def ESPIPE(): Int { return 29; }
def EROFS(): Int { return 30; }
def EMLINK(): Int { return 31; }
def EPIPE(): Int { return 32; }
def EDOM(): Int { return 33; }
def ERANGE(): Int { return 34; }

def description(err: Int): *Byte {
    if err == EPERM() {
        return b"EPERM: Operation not permitted";
    }
    if err == ENOENT() {
        return b"ENOENT: No such file or directory";
    }
    if err == ESRCH() {
        return b"ESRCH: No such process";
    }
    if err == EINTR() {
        return b"EINTR: Interrupted system call";
    }
    if err == EIO() {
        return b"EIO: I/O error";
    }
    if err == ENXIO() {
        return b"ENXIO: No such device or address";
    }
    if err == E2BIG() {
        return b"E2BIG: Argument list too long";
    }
    if err == ENOEXEC() {
        return b"ENOEXEC: Exec format error";
    }
    if err == EBADF() {
        return b"EBADF: Bad file number";
    }
    if err == ECHILD() {
        return b"ECHILD: No child processes";
    }
    if err == EAGAIN() {
        return b"EAGAIN: Try again";
    }
    if err == ENOMEM() {
        return b"ENOMEM: Out of memory";
    }
    if err == EACCES() {
        return b"EACCES: Permission denied";
    }
    if err == EFAULT() {
        return b"EFAULT: Bad address";
    }
    if err == ENOTBLK() {
        return b"ENOTBLK: Block device required";
    }
    if err == EBUSY() {
        return b"EBUSY: Device or resource busy";
    }
    if err == EEXIST() {
        return b"EEXIST: File exists";
    }
    if err == EXDEV() {
        return b"EXDEV: Cross-device link";
    }
    if err == ENODEV() {
        return b"ENODEV: No such device";
    }
    if err == ENOTDIR() {
        return b"ENOTDIR: Not a directory";
    }
    if err == EISDIR() {
        return b"EISDIR: Is a directory";
    }
    if err == EINVAL() {
        return b"EINVAL: Invalid argument";
    }
    if err == ENFILE() {
        return b"ENFILE: File table overflow";
    }
    if err == EMFILE() {
        return b"EMFILE: Too many open files";
    }
    if err == ENOTTY() {
        return b"ENOTTY: Not a typewriter";
    }
    if err == ETXTBSY() {
        return b"ETXTBSY: Text file busy";
    }
    if err == EFBIG() {
        return b"EFBIG: File too large";
    }
    if err == ENOSPC() {
        return b"ENOSPC: No space left on device";
    }
    if err == ESPIPE() {
        return b"ESPIPE: Illegal seek";
    }
    if err == EROFS() {
        return b"EROFS: Read-only file system";
    }
    if err == EMLINK() {
        return b"EMLINK: Too many links";
    }
    if err == EPIPE() {
        return b"EPIPE: Broken pipe";
    }
    if err == EDOM() {
        return b"EDOM: Math argument out of domain of func";
    }
    if err == ERANGE() {
        return b"ERANGE: Math result not representable";
    }
    return b"";
}