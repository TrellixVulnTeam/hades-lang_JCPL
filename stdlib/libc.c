#include <dirent.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <stdio.h>
#include <errno.h>

const char* hdc_dirent_name(struct dirent* entry) {
    return entry->d_name;
}

unsigned char hdc_dirent_type(struct dirent* entry)  {
    return entry->d_type;
}

unsigned char hdc_dirent_is_directory(struct dirent* entry) {
    return S_ISDIR(entry->d_type);
}

void hdc_set_errno(int value) {
    errno = value;
}


FILE* hdc_get_stdout() {
    return stdout;
}


int hdc_errno() {
    return errno;
}