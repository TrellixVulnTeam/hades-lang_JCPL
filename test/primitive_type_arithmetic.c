#include <stdint.h>
#include <stddef.h>
#include <stdio.h>

void pta_print_u8(uint8_t value) {
    printf("%hhu", value);
}

void pta_print_i8(int8_t value) {
    printf("%hhi", value);
}

void pta_print_i16(int16_t value) {
    printf("%hi", value);
}

void pta_print_u16(uint16_t value) {
    printf("%hu", value);
}

void pta_print_u32(uint32_t value) {
    printf("%u", value);
}

void pta_print_i32(int32_t value) {
    printf("%i", value);
}

void pta_print_i64(int64_t value) {
    printf("%li", value);
}

void pta_print_u64(uint64_t value) {
    printf("%lu", value);
}