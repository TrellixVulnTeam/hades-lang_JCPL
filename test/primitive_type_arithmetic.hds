import libc as c;

def main(): Void {
    print_i8(0); line();
    print_i8(-1); line();
    print_i8(1); line();
    print_i8(63); line();
    print_i8(-64); line();
    print_i8(1 - 2); line();

    print_u8(127); line();
    print_u8(0); line();
}

def line(): Void {
    c.puts(b"");
}

extern def print_i8(i8): Void = pta_print_i8;
extern def print_u8(u8): Void = pta_print_u8;

extern def print_i16(i16): Void = pta_print_i16;
extern def print_u16(u16): Void = pta_print_u16;

extern def print_i32(i32): Void = pta_print_i32;
extern def print_u32(u32): Void = pta_print_u32;

extern def print_i64(i64): Void = pta_print_64;
extern def print_u64(u64): Void = pta_print_64;

