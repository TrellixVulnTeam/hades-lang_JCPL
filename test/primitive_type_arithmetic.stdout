0
-1
1
63
-64
-1
127
0